File: code/Kotlin/libs/annotation_processing/apt/compiler/src/main/java/inject/processor/creator/ServiceMethod.java
package inject.processor.creator;

import static javax.lang.model.element.Modifier.PUBLIC;
import static javax.lang.model.element.Modifier.STATIC;

import com.squareup.javapoet.MethodSpec;

import javax.lang.model.element.Element;

import inject.classname.AndroidClassName;
import inject.kotlin.KotlinClassName;

/**
 * @author yuansui
 * @since 2020/9/8
 */
class ServiceMethod extends BaseMethod {

    ServiceMethod(CreatorProcessor processor, Element annotatedElement) {
        super(processor, annotatedElement);
    }

    MethodSpec start() {
        MethodSpec.Builder builder = MethodSpec.methodBuilder("start")
                .addModifiers(PUBLIC)
                .addParameter(processor.createNullableContext())
                .addStatement("if(context == null) return");

        // 直接调用之前生成的newIntent()方法
        builder.addStatement(processor.getIntentParamsText(allFields), AndroidClassName.INTENT);

        // 直接调用Launcher, 减少生成的代码量
        builder.addStatement("$T.startService(context, intent, null)", KotlinClassName.LAUNCHER);

        return builder.build();
    }

    MethodSpec startForeground() {
        MethodSpec.Builder builder = MethodSpec.methodBuilder("startForeground")
                .addModifiers(PUBLIC)
                .addParameter(processor.createNullableContext())
                .addStatement("if(context == null) return");

        // 直接调用之前生成的newIntent()方法
        builder.addStatement(processor.getIntentParamsText(allFields), AndroidClassName.INTENT);

        // 直接调用Launcher, 减少生成的代码量
        builder.addStatement("$T.startForegroundService(context, intent, null)", KotlinClassName.LAUNCHER);

        return builder.build();
    }

    MethodSpec stop() {
        return MethodSpec.methodBuilder("stop")
                .addModifiers(PUBLIC, STATIC)
                .addParameter(processor.createNullableContext())
                .addStatement("if(context == null) return")
                .addStatement("$T intent = new Intent(context, $T.class)", AndroidClassName.INTENT, annotatedTypeName)
                .addStatement("context.stopService(intent)")
                .build();
    }

}

File: code/Kotlin/libs/annotation_processing/kapt/compiler/src/main/java/inject/compiler/method/creator/ServiceMethod.kt
package inject.compiler.method.creator

import com.squareup.kotlinpoet.FunSpec
import inject.compiler.classname.AndroidClassName
import inject.compiler.classname.VectorClassName
import inject.compiler.method.creator.BaseMethod
import inject.compiler.ext.createNullableContext
import inject.compiler.processor.FormatSpecifiers
import inject.compiler.processor.CreatorProcessor
import javax.lang.model.element.Element

/**
 * @author yuansui
 * @since 2020/9/8
 */
internal class ServiceMethod(
    processor: CreatorProcessor,
    annotatedElement: Element
) : BaseMethod<CreatorProcessor>(processor, annotatedElement) {

    fun start(): FunSpec {
        val builder = FunSpec.builder("start")
            .addParameter(createNullableContext())
            .addStatement("if(context == null) return")

        // 直接调用之前生成的newIntent()方法
        builder.addStatement(processor.getIntentParamsText(allFields), AndroidClassName.INTENT)

        // 直接调用Launcher, 减少生成的代码量
        builder.addStatement(
            "${FormatSpecifiers.TYPE}.startService(context, intent, null)",
            VectorClassName.LAUNCHER
        )

        return builder.build()
    }

    fun startForeground(): FunSpec {
        val builder: FunSpec.Builder = FunSpec.builder("startForeground")
            .addParameter(createNullableContext())
            .addStatement("if(context == null) return")

        // 直接调用之前生成的newIntent()方法
        builder.addStatement(processor.getIntentParamsText(allFields), AndroidClassName.INTENT)

        // 直接调用Launcher, 减少生成的代码量
        builder.addStatement(
            "${FormatSpecifiers.TYPE}.startForegroundService(context, intent, null)",
            VectorClassName.LAUNCHER
        )

        return builder.build()
    }

    // TODO: 暂时不添加stop方法, 因为只能有一个companion对象, 需要修改Processor的写法, 目前看来不太需要
//    fun stop(): TypeSpec {
//        val builder = TypeSpec.companionObjectBuilder()
//        builder.addFunction(
//            FunSpec.builder("stop")
//                .addParameter(createNullableContext())
//                .addStatement("if(context == null) return")
//                .addStatement(
//                    "val intent = Intent(context, ${FormatSpecifiers.NAME}.class)",
//                    annotatedTypeName
//                )
//                .addStatement("context.stopService(intent)")
//                .build()
//        )
//        return builder.build()
//    }
}
